arbnode/transaction_streamer.go

```diff
@@ -192,42 +212,69 @@ func deleteStartingAt(db ethdb.Database, batch ethdb.Batch, prefix []byte, minKe
 	return iter.Error()
 }
 
-func (s *TransactionStreamer) reorgToInternal(batch ethdb.Batch, count arbutil.MessageIndex) error {
+// The insertion mutex must be held. This acquires the reorg mutex.
+// Note: oldMessages will be empty if reorgHook is nil
+func (s *TransactionStreamer) reorgToInternal(batch ethdb.Batch, count arbutil.MessageIndex) ([]*arbstate.MessageWithMetadata, error) {
 	if count == 0 {
-		return errors.New("cannot reorg out init message")
+		return nil, errors.New("cannot reorg out init message")
+	}
+	var oldMessages []*arbstate.MessageWithMetadata
+	if s.reorgSequencing != nil {
+		targetMsgCount, err := s.GetMessageCount()
+		if err != nil {
+			return nil, err
+		}
+		config := s.config()
+		maxResequenceMsgCount := count + arbutil.MessageIndex(config.MaxReorgResequenceDepth)
+		if config.MaxReorgResequenceDepth >= 0 && maxResequenceMsgCount < targetMsgCount {
+			log.Error(
+				"unable to re-sequence all old messages because there are too many",
+				"reorgingToCount", count,
+				"removingMessages", targetMsgCount-count,
+				"maxReorgResequenceDepth", config.MaxReorgResequenceDepth,
+			)
+			targetMsgCount = maxResequenceMsgCount
+		}
+		for i := count; i < targetMsgCount; i++ {
+			oldMessage, err := s.GetMessage(i)
+			if err != nil {
+				return nil, err
```

Comment:
maybe log the error and break instead?
We probably want to allow a non-sequencer to reorg overwriting bad messages in the database.